crates/bevy_ecs/src/query/state.rs

```diff
@@ -1276,43 +1339,156 @@ impl<Q: WorldQuery, F: ReadOnlyWorldQuery> QueryState<Q, F> {
+            for component_id in archetype.components() {
+                if query_without_components.contains(component_id.index()) {
+                    query_without_components_in_entity.push((
+                        components.get_info(component_id).unwrap().name().to_owned(),
+                        component_id,
+                    ));
+                } else if query_with_components.contains(component_id.index()) {
+                    query_unmatched_components.set(component_id.index(), false);
+                }
+            }
+
+            let query_with_components_not_in_entity = query_unmatched_components
+                .ones()
+                .map(|component_id| {
+                    let component_id = ComponentId::new(component_id);
+                    (
+                        components.get_info(component_id).unwrap().name().to_owned(),
+                        component_id,
+                    )
+                })
+                .collect();
+
+            mismatch_details.push(QueryEntityComponentMismatch {
+                query_with_components_not_in_entity,
+                query_without_components_in_entity,
+                entity_archetype: entity_location.archetype_id,
+                query_with_components,
+                query_without_components,
+            });
+        }
+
+        mismatch_details
+    }
 }
 
 /// An error that occurs when retrieving a specific [`Entity`]'s query result from [`Query`](crate::system::Query) or [`QueryState`].
-// TODO: return the type_name as part of this error
-#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[derive(Debug, PartialEq, Eq, Clone)]
 pub enum QueryEntityError {
-    /// The given [`Entity`]'s components do not match the query.
-    ///
-    /// Either it does not have a requested component, or it has a component which the query filters out.
-    QueryDoesNotMatch(Entity),
+    /// The given [`Entity`]'s components do not match the query, see [`QueryEntityMismatchDetail`].
+    QueryDoesNotMatch(QueryEntityMismatchDetail, QueryEntityErrorDetail),
     /// The given [`Entity`] does not exist.
-    NoSuchEntity(Entity),
+    NoSuchEntity(QueryEntityErrorDetail),
     /// The [`Entity`] was requested mutably more than once.
     ///
     /// See [`QueryState::get_many_mut`] for an example.
-    AliasedMutability(Entity),
+    AliasedMutability(QueryEntityErrorDetail),
+}
+
+/// Additional information in the context of a [`QueryEntityError`].
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct QueryEntityErrorDetail {
+    /// Specific entity which was requested when encountering the error.
+    pub requested_entity: Entity,
+    /// Representation of the query's type.
+    pub query_type: &'static str,
+}
+
+/// Diagnosis about why an [`Entity`] and a [`Query`](crate::system::Query) or [`QueryState`] don't match.
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum QueryEntityMismatchDetail {
```

Comment:
I think the `Detail` here only contributes to make the type harder to read. I'd rather see a shorter name.

```suggestion
pub enum QueryEntityMismatch {
```